### PR TITLE
feat: Use tensorflow-datasets (tfds) for dataset operations

### PR DESCRIPTION
**Pull request by streino**
This PR replaces the dual use of pandas DataFrame + jsonl generator-based td.data.Dataset with a tfds dataset.

What we gain:
- We never load the full data in RAM, so we have less risks of running into memory issues when we'll add features.
- tfds stores data in the TFRecord format (and handles the serde for us). It's supposed to be optimal for the TF use-case. I don't see any difference in training times (didn't profile yet to see if this can be improved), but it's very fast to iterate over the data (vs jsonl generator). So now we can iterate as much as we want during preprocessing and replace pandas operations. It's obviously slower than pandas on in-memory data but still fast enough.
- tfds handles downloading and caching the raw data files (jsonl) as well as caching the tfds dataset. So we can get rid of download.sh.
- tfds is designed for publicly shared datasets. So if at some point Open Food Fact wants to release its dataset, it should be easy to do so.
- It adds doc (to be completed) and a schema to the data. Should make it easier to approach than a bunch of untyped/unnamed tensors.

I also added a `Train-tfds-refactor-example.ipynb` notebook:
- It's a simplified version of the first part of `Train.ipynb` (up to training step) to show how tfds is integrated. It doesn't use the `category_classification` lib so all tfds-related operations are visible in the notebook.
- Second part of the notebook is a proposal to build the model "per feature". I think it's easier to understand and tweak with new features/models. It's similar to what @8huit is doing, so I'm putting it up here for discussion.

### CODE REVIEW

**Comment by streino** [comment]
Leaving as draft while I fix the few remaning FIXMEs.

File: datasets/off_category/off_category.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  'product_name': Feature(tfds.features.Tensor(shape=(), dtype=tf.string), ''),
+  'ingredients_tags': Feature(tfds.features.Tensor(shape=(None,), dtype=tf.string), [],
+                              input_field='known_ingredient_tags'),
+  'categories_tags': Feature(tfds.features.Tensor(shape=(None,), dtype=tf.string), [])
+}
+
+_LABEL = 'categories_tags'
+
+
+class OffCategory(tfds.core.GeneratorBasedBuilder):
```
Comment (streino) [inline datasets/off_category/off_category.py:40]: Should we call the dataset 'off_category' or 'off_categories'? Repo is 'category-classification' but products have multiple categories. Not sure we care but since tfds manages datasets by name, it'll be more painful to change it later.

File: datasets/off_category/off_category.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  '0.0.1': 'Initial test'
+}
+
+# Don't forget to run `tfds build --register_checksums` when changing the data source
+_DATA_URL = 'https://github.com/openfoodfacts/openfoodfacts-ai/releases/download/dataset-category-2021-09-15/category_xx.{split}.jsonl.gz'
+
+_FEATURES = {
+  'code': Feature(tfds.features.Tensor(shape=(), dtype=tf.string), ''),
+  'product_name': Feature(tfds.features.Tensor(shape=(), dtype=tf.string), ''),
+  'ingredients_tags': Feature(tfds.features.Tensor(shape=(None,), dtype=tf.string), [],
```
Comment (streino) [inline datasets/off_category/off_category.py:32]: I'm chose to decorrelate data fields and feature names, because:
- Some future features won't exactly match data fields anyway (e.g. for nutriments we'll extract some sub-fields as independent features, and probably have a catch-all feature with the rest of the sub-fields, like we discussed with @alexgarel a few meetings ago).
- Model Inputs are based on the feature name, so we'd have to update the Inputs names when changing target fields. For instance, when updating to the new dataset, we'll go from 'known_ingredients_tags' to 'ingredients_tags'. If we decorrelate field/feature names like here we won't have to do anything model-side. Otherwise we'll have to update the model. Maybe it's better since it's explicit, or it's just more cumbersome.

Any thoughts about this?

Comment (8huit) [reply datasets/off_category/off_category.py:32]: This PR si a great addition to improve scalability of the model!
I think decorrelating field / feature will enhance the reusability of the model ;)

**Comment by streino** [comment]
The more I refactor/polish, the more I like the "per feature" proposal in the second part of the `Train-tfds-refactor-example.ipynb` notebook :)

I find it more readable as it keeps related stuff closer together, and eliminates the need for global vars like `FEATURES` (which we build along the way, like the model `inputs`).

I think we should go ahead and adopt it in `Train.ipynb`. Does that sound good to everyone? @8huit @alexgarel? If yes, I'll go ahead and refactor `Train.ipynb` in this PR.

**Comment by 8huit** [comment]
I agree with the "per feature" organization, to foster new feature experimentation if needed.
And yes, gathering related pieces together make it easier to read and to understand ;)

File: datasets/off_category/off_category.py (hunk 1 above)
Comment (alexgarel) [reply datasets/off_category/off_category.py:40]: I think off_categories is better indeed.

**Comment by streino** [comment]
PR is finally ready for review :)

Some additional changes (vs what I wrote in the PR description):
- Each trained model is now stored in a `model.{timestamp}` dir, with the timestamp set in the "training" cell of the notebook (the one calling `model.fit`). Before, all runs would land in the `models` directory, so if you forgot to clear it up before launching another run, you'd get several models jumbled up together. Note: Don't forget to clean obsolete `model.{timestamp}` once in a while...
- Test predictions are done with the "serving" model so we get the categories tag names for free, instead of having to map them manually in the notebook. Plus it exercises the "serving" model, which is what we'll be deploying, so better make sure it works. I kept the "training" model predictions in the notebook but I'm not sure it's needed?
- We used to download the categories taxonomy and copy it in the saved model. We rely on TFDS to download dataset files so I thought we could retire `download.sh`. But if we still want to download the taxonomy (and the agribalyse file), we should probably keep `download.sh` for non-dataset assets.
- Since we went ahead with TFDS in `Train.ipynb`, I removed the `Train-tfds-refactor-example.ipynb` file from the PR.

File: experiments/Train.ipynb [hunk 3]
```diff
@@ -159,380 +150,840 @@
...
+     },
+     "execution_count": 8,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "%%time\n",
+    "\n",
+    "feature_name = 'product_name'\n",
```
Comment (alexgarel) [inline experiments/Train.ipynb:309]: It would be cool to comment more this kind of cells in the future. What are we doing, why we try this, etc.

File: experiments/Train.ipynb [hunk 4]
```diff
@@ -159,380 +150,840 @@
...
+    }
+   ],
+   "source": [
+    "%%time\n",
+    "\n",
+    "labels = 'categories_tags'\n",
+    "\n",
+    "categories_vocab = get_vocabulary(\n",
+    "    flat_batch(select_feature(ds, labels), batch_size=PREPROC_BATCH_SIZE),\n",
+    "    min_freq = model_config.category_min_count)\n",
```
Comment (alexgarel) [inline experiments/Train.ipynb:425]: Hum, this parameter should have a comment. I think it can really change a lot performances.
If I wanted to cheat, I would make this parameter up ;-)

**Review by alexgarel — APPROVED** [review]
Really cool PR. Thanks @streino.

On my side we can merge ASAP.

File: datasets/off_categories/off_categories.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  VERSION = tfds.core.Version('1.0.0')
+  RELEASE_NOTES = _RELEASE_NOTES
+
+  def _info(self) -> tfds.core.DatasetInfo:
+    return tfds.core.DatasetInfo(
+        builder=self,
+        description=_DESCRIPTION,
+        features=tfds.features.FeaturesDict({k: f.spec for k, f in _FEATURES.items()}),
+        supervised_keys=({k: k for k in _FEATURES.keys() if k != _LABEL}, _LABEL),
+        disable_shuffling=True,
```
[the remark this answers is not in the document]
Comment (streino) [reply datasets/off_categories/off_categories.py:50]: This is to keep the same order as with the original json generator-based dataset, so we could compare the results before/after this PR.

In the future (e.g. when we switch to the new dataset) I think we should enable shuffling, here and also per-epoch during training. This should make for more robust models.

But before we do the switch, I'd like to verify that `disable_shuffling=False` is deterministic and will build the same dataset on all machines that call `builder.download_and_prepare()`.

File: experiments/Train.ipynb (hunk 3 above)
Comment (streino) [reply experiments/Train.ipynb:309]: I'll pass for this PR in the interest of time, but good point. If someone wants to tackle that part, please go ahead!!

File: category_classification/data_utils.py [hunk 6]
```diff
@@ -1,90 +1,134 @@
...
+    list
+        List of tokens in the vocabulary
+    """
+    counter = Counter()
+    for batch in ds:
+        counter.update(batch.numpy())
+    return [
+        x[0].decode()
+        for x in itertools.takewhile(lambda x: x[1] >= min_freq, counter.most_common())
+    ]
```
[the remark this answers is not in the document]
Comment (streino) [reply category_classification/data_utils.py:123]: It's nice to be able to rely on the included batteries :)

**Review by 8huit — APPROVED** [review]
great PR! ;)
